doc/api/async_hooks.md

```diff
@@ -804,6 +804,162 @@ for (let i = 0; i < 10; i++) {
 }
 ```
 
+## Class: AsyncLocal
+<!-- YAML
+added: REPLACEME
+-->
+
+This class is used to create asynchronous state between callbacks and promise
+chains. It allows you to keep variables throughout the lifetime of a web request
+or any other asynchronous duration. It is similar to thread local storage
+ in other languages.
+
+The following example builds a logger that will always know the current HTTP
+request and use it to display enhanced logs without needing to explicitly pass
+the current HTTP request to it.
+
+```js
+const { AsyncLocal } = require('async_hooks');
+const http = require('http');
+
+const kReq = 'CURRENT_REQUEST';
+const asyncLocal = new AsyncLocal();
+
+function log(...args) {
+  const store = asyncLocal.getStore();
+  // Make sure the store exists and it contains a request.
+  if (store && store.has(kReq)) {
+    const req = store.get(kReq);
+    // Prints `GET /items ERR could not do something
+    console.log(req.method, req.url, ...args);
+  } else {
+    console.log(...args);
+  }
+}
+
+http.createServer((request, response) => {
+  asyncLocal.run(() => {
+    const store = asyncLocal.getStore();
+    store.set(kReq, request);
+    someAsyncOperation((err, result) => {
+      if (err) {
+        log('ERR', err.message);
+      }
+    });
+  });
+})
+.listen(8080);
+```
+
+When having multiple instance of `AsyncLocal`, they are independant from each
+others. It is safe to instanciate this class multiple times.
+
+### Choosing between asynchronous and synchronous methods
+
+AsyncLocal offer three pair of methods to enter and exit a context:
+* `run/exit`
+* `runAndReturn/exitAndReturn`
+
+#### When to choose `run/exit`
+These methods are asynchronous. They are called with a callback function that
+runs within a new asynchronous call. This is the most explicit behavior as
+everything that is executed within the callback of `run` (including further
+asynchronous operations) will have access to the store. The `exit` method has
+a similar behavior to stop context propagation.
+
+If an instance of AsyncLocal is used for error management (for instance with
+`process.setUncaughtExceptionCaptureCallback`). Only the exceptions thrown
+in the scope of the callback function will be associated with the context.
+
+These methods are the safest to call as they provide strong scoping.
+
+#### When to choose `runAndReturn/exitAndReturn`
+These methods are synchronous and scoped. The callback function will be executed
+synchronously and it's return value will be returned by `runAndReturn`
+and `exitAndReturn`. The store will only be accessible from within the callback
+function and the asynchronous operations created within this scope.
+If the callback throws an error, `runAndReturn` and `exitAndReturn` will throw
+it too.
+
+These methods provide good scoping while being synchronous.
+
```

Comment:
From reading the documentation, it's not entirely clear to me how to use run & exit API in code base using purely Promises, `async` functions and `await` keyword.

Consider a [Koa](https://www.npmjs.com/package/koa) middleware setting up a per-request async storage:

```js
app.use(async (ctx, next) => {
  // create a new scope
  await next();
  // exit the scope
});
```

My first try would be something along the following lines. It looks rather hacky to me, will it even work as intended?

```js
const { AsyncLocal } = require('async_hooks');
const { promisify } = require('util');

const asyncLocal = new AsyncLocal();
asyncLocal.runAsync = promisify(asyncLocal.run);
asyncLocal.exitAsync = promisify(asyncLocal.exit);

const Koa = require('koa');
const app = new Koa();

app.use(async setupClsScope(ctx, next) => {
  await asyncLocal.runAsync();
  await next();
  await asyncLocal.exitAsync();
});
```

Also: what is the interaction of this new feature (async storage) with zero-cost async stack traces provided by V8 for async/await based code? (See https://v8.dev/blog/fast-async#improved-developer-experience). In my example code shown above, when an error is thrown from `exit`, the stack trace should include `setupClsScope` frame. How is this going to work with the callback-based implementation?

Reply:
That's two great questions:
-> zero-cost async stack traces: in theory this has no impact as this is an orthogonal topic (on me to check that deeper)

-> Koa use case:
I have experimented a couple things and I need to dig deeper before giving you a precise answer. @Flarna, @Qard , given the experimentations, I think the easiest way to cover this with the current system is to re-enable `enterSync/`exitSync` and to call them before and after `await next()`. But I am not sure I have looked hard enough into the issue yet. I had a packed Thursday, hopefully Friday and/or Saturday will be easier on me.

Reply:
I fear that `run`/`exit` are not nice regarding zero-cost async stack traces as they use `process.nextTick` internally but  zero-cost async stack traces work only from `await` to `await`.

Reply:
Regarding the koa use case I think that the following should work:
```js
app.use(async (ctx, next) => {
  await asyncLocal.runAndReturn(next);
});
```
This should be also fine regarding zero-cost async stack traces as `runAndReturn` as a sync function directly calling `next` and forwards return value.

Reply:
I have played around with 
I guess a solution would be to allow no callback to be passed to `runAndReturn/exitAndReturn`, it would give us the following cases:

```js
localStorage.runAndReturn();
localStorage.getStore() === undefined // tre

await localStorage.runAndReturn();
localStorage.getStore() // storage available
await localStorage.exitAndReturn();
``` 
In the second case, the store is still scoped to the promise chain. In that case, `runAndReturn/exitAndReturn` are not proper name anymore however.

Reply:
@Flarna this is correct. The noticeable thing is that the context is still available in the rest of the promise chain (which is the logical behavior):
```js
app.use(async (ctx, next) => {
  await asyncLocal.runAndReturn(next);
  asyncLocal.getStore() // returns the store
});
```

This can be exited with
```js
await asyncLocal.exitAndReturn(() => Promise.resolve());
```

Reply:
I will update the doc to describe this.

Reply:
> ```js
> app.use(async (ctx, next) => {
>   await asyncLocal.runAndReturn(next);
>   asyncLocal.getStore() // returns the store
> });
> ```

If this is true then `runAndReturn` has a bug as it should exit the scope before returning.

Reply:
> I will update the doc to describe this.

Awesome, such documentation would be super helpful.

Do we have any advice for people not using Koa-style `next`? For example:

```ts
#!/usr/bin/env node

main().catch(err => {
  console.error('Script failed.', err);
  process.exit(1);
});

async function main() {
  // TBD: create a new scope
  const data = await doSomeStuff();
  const result = await doSomethingElse(data);
  console.log(result);
  // TBD: exit the scope (maybe not needed)?
}
```